src/components/calcite-modal/calcite-modal.tsx

```diff
@@ -0,0 +1,195 @@
+import {
+  Component,
+  Element,
+  Prop,
+  Host,
+  Event,
+  EventEmitter,
+  Listen,
+  h,
+  Method,
+  State
+} from "@stencil/core";
+import { x24 } from "@esri/calcite-ui-icons";
+import "@a11y/focus-trap";
+import { FocusTrap } from "@a11y/focus-trap";
+import { getElementDir, getElementTheme, hasSlottedContent } from "../../utils/dom";
+
+@Component({
+  tag: "calcite-modal",
+  styleUrl: "calcite-modal.scss",
+  shadow: true
+})
+export class CalciteModal {
+  //--------------------------------------------------------------------------
+  //
+  //  Element
+  //
+  //--------------------------------------------------------------------------
+  @Element() el: HTMLElement;
+
+  //--------------------------------------------------------------------------
+  //
+  //  Properties
+  //
+  //--------------------------------------------------------------------------
+  /** Optionally pass a function to run before close */
+  @Prop() beforeClose: (el: HTMLElement) => Promise<void> = () =>
+    Promise.resolve();
+  /** Aria label for the close button */
+  @Prop() closeLabel: string = "Close";
+  /** Prevent the modal from taking up the entire screen on mobile */
+  @Prop({ reflect: true }) docked: boolean;
+  /** Specify an element to focus when the modal is first opened */
+  @Prop() firstFocus?: HTMLElement;
+  /** Set the overall size of the modal */
+  @Prop({ reflect: true }) size: "small" | "medium" | "large" | "fullscreen" =
+    "small";
+  /** Adds a color bar at the top for visual impact,
+   * Use color to add importance to desctructive/workflow dialogs. */
+  @Prop({ reflect: true }) color?: "red" | "blue";
+  /** Select theme (light or dark) */
+  @Prop({ reflect: true })
+  theme: "light" | "dark" = "light";
+
+  //--------------------------------------------------------------------------
+  //
+  //  Lifecycle
+  //
+  //--------------------------------------------------------------------------
+  componentDidLoad() {
+    const back = this.el.shadowRoot.querySelector(
+      "slot[name=back]"
+    ) as HTMLSlotElement;
+    const secondary = this.el.shadowRoot.querySelector(
+      "slot[name=secondary]"
+    ) as HTMLSlotElement;
+    back.addEventListener("slotchange", () => {
+      this.hideBackButton = !hasSlottedContent(back);
+    });
+    secondary.addEventListener("slotchange", () => {
+      this.hideSecondaryButton = !hasSlottedContent(secondary);
+    });
+  }
+
+  render() {
+    const dir = getElementDir(this.el);
+    const theme = getElementTheme(this.el);
+    return (
+      <Host
+        role="dialog"
+        aria-modal="true"
+        class={{ "is-active": this.isActive }}
```

Comment:
I know we are all over the place with how we've implemented this, but is there a reason for using a class vs an "active"/"is-active" attribute here?

Reply:
Yeah, my thinking was if it's an attribute we run the risk of consuming devs setting that property and thinking they're opening the modal, but missing all the event listeners/a11y stuff.